src/script/interpreter.cpp

```diff
@@ -1264,6 +1288,109 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                 }
                 break;
 
+                case OP_VAULT_RECOVER:
+                {
+                    // Only available in Tapscript
+                    if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0) {
+                        return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+                    }
+
+                    // Stack:
+                    //  - <recovery-sPK-hash>
+                    //  - <recovery-vout-idx>
+                    if (stack.size() < 2) {
+                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+                    }
+
+                    const valtype& hash_bytes_from_stack = stacktop(-1);
+                    if (hash_bytes_from_stack.size() != 32) {
+                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+                    }
+                    const uint256 recovery_spk_hash{hash_bytes_from_stack};
+
+                    std::optional<size_t> recovery_vout_idx;
+                    if (!GetVoutIdxFromStack(recovery_vout_idx, -2) || !recovery_vout_idx) {
+                        return set_error(serror, SCRIPT_ERR_VAULT_BAD_VOUT_IDX);
+                    }
+
+                    if (const auto& err = checker.CheckVaultRecover(
+                            *recovery_vout_idx, execdata, recovery_spk_hash, script, flags)) {
+                        return set_error(serror, *err);
+                    }
+
+                    popstack(stack);
+                    popstack(stack);
+                    stack.push_back(vchTrue);
+                }
+                break;
+
+                case OP_VAULT:
+                {
+                    // Only available in Tapscript
+                    if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0) {
```

Comment:
This should fail anytime it's not tapscript specifically, which then allows you to clean up `CheckVaultTrigger` significantly, as you know precisely which versions you're dealing with (including tapleaf version). 

A new output version or tapleaf version is likely to have very different semantics anyways (aka script replacements)

Reply:
Good point, fixing this (cc @darosior).